src/row_codec.rs

```diff
@@ -1,182 +1,219 @@
+/// | `expire_ts`      | `u64` | Optional, only has value when flags & HAS_EXPIRE_TS    |
+/// | `create_ts`      | `u64` | Optional, only has value when flags & HAS_CREATE_TS    |
+/// | `value_len`      | `u32` | Length of the value                                    |
+/// | `value`          | `var` | Value bytes                                            |
+
+pub(crate) struct SstRowEntry {
     key_prefix_len: usize,
-    key_suffix: &[u8],
-    value: Option<&[u8]>,
-    row_features: &[RowFeature],
-    timestamp: Option<i64>,
-    expire_at: Option<i64>,
-) {
-    data.put_u16(key_prefix_len as u16);
-    data.put_u16(key_suffix.len() as u16);
-    data.put(key_suffix);
-
-    data.put(encode_meta(
-        row_features,
-        value.is_none(),
-        timestamp,
-        expire_at,
-    ));
-
-    if let Some(value) = value {
-        data.put_u32(value.len() as u32);
-        data.put(value);
-    }
+    key_suffix: Bytes,
+    seq: u64,
+    expire_ts: Option<i64>,
+    create_ts: Option<i64>,
+    value: ValueDeletable,
 }
 
-fn encode_meta(
-    row_features: &[RowFeature],
-    is_tombstone: bool,
-    timestamp: Option<i64>,
-    expire_at: Option<i64>,
-) -> Bytes {
-    let mut meta = BytesMut::new();
-
-    for attr in row_features {
-        match attr {
-            RowFeature::Flags => meta.put_u8(encode_row_flags(is_tombstone)),
-            RowFeature::Timestamp => meta.put_i64(timestamp.expect("Timestamp RowAttribute was enabled for SST but attempted to insert a row with no timestamp.")),
-            RowFeature::ExpireAtTs => meta.put_i64(expire_at.unwrap_or(NO_EXPIRE_TS))
+impl SstRowEntry {
+    pub fn new(
+        key_prefix_len: usize,
+        key_suffix: Bytes,
+        seq: u64,
+        value: ValueDeletable,
+        create_ts: Option<i64>,
+        expire_ts: Option<i64>,
+    ) -> Self {
+        Self {
+            key_prefix_len,
+            key_suffix,
+            seq,
+            create_ts,
+            expire_ts,
+            value,
         }
     }
 
-    meta.freeze()
-}
-
-fn encode_row_flags(is_tombstone: bool) -> u8 {
-    let mut flags = RowFlags::empty();
-
-    if is_tombstone {
-        flags |= RowFlags::Tombstone;
+    fn flags(&self) -> RowFlags {
+        let mut flags = match &self.value {
+            ValueDeletable::Value(_) => RowFlags::default(),
+            ValueDeletable::Tombstone => RowFlags::TOMBSTONE,
+        };
+        if self.expire_ts.is_some() {
+            flags |= RowFlags::HAS_EXPIRE_TS;
+        }
+        if self.create_ts.is_some() {
+            flags |= RowFlags::HAS_CREATE_TS;
+        }
+        flags
     }
 
-    flags.bits()
+    // only called after decoding from SST file, the full key should be reconstructed with prefix
+    // concatenated with suffix. when the full key is constructed, the prefix length is zero, and
+    // the suffix is the full key.
+    fn key(&self) -> &Bytes {
```

Comment:
It seems like we only use this from `into`. Instead of using `Into<RowEntry>` why not just have this be something like `build_key_using_prefix(&self, first_key: &Bytes) -> Bytes` that constructs the key from the first key. This way we don't need the `unreachable` assertion.

Reply:
love this idea, let me change it  :+1:

Reply:
made a refactor on this in e724e4e